crates/sdk/src/signing.rs

```diff
@@ -2030,3 +2034,925 @@ pub async fn to_ledger_vector(
+    use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+    use tokio::sync::{Mutex, RwLockReadGuard, RwLockWriteGuard};
+
+    use super::*;
+    use crate::args::InputAmount;
+    use crate::masp::fs::FsShieldedUtils;
+    use crate::masp::{ShieldedContext, WalletMap};
+
+    fn arbitrary_args() -> args::Tx {
+        args::Tx {
+            dry_run: false,
+            dry_run_wrapper: false,
+            dump_tx: false,
+            output_folder: None,
+            force: false,
+            broadcast_only: false,
+            ledger_address: tendermint_rpc::Url::from_str(
+                "http://127.0.0.1:42",
+            )
+            .expect("Test failed"),
+            initialized_account_alias: None,
+            wallet_alias_force: false,
+            fee_amount: None,
+            wrapper_fee_payer: None,
+            fee_token: Address::Internal(InternalAddress::Governance),
+            gas_limit: namada_tx::data::GasLimit::from(2),
+            expiration: Default::default(),
+            chain_id: None,
+            signing_keys: vec![],
+            signatures: vec![],
+            tx_reveal_code_path: Default::default(),
+            password: Some(zeroize::Zeroizing::new("bingbong123".to_string())),
+            memo: None,
+            use_device: false,
+            device_transport: Default::default(),
+        }
+    }
+
+    pub struct TestNamadaImpl {
+        wallet: RwLock<Wallet<TestWalletUtils>>,
+        client: TestClient,
+        io: StdIo,
+    }
+
+    impl TestNamadaImpl {
+        fn new(
+            paths: Option<HashSet<String>>,
+        ) -> (Self, UnboundedSender<Option<EncodedResponseQuery>>) {
+            let (send, recv) = tokio::sync::mpsc::unbounded_channel();
+            (
+                Self {
+                    wallet: RwLock::new(Wallet::new(
+                        TestWalletUtils,
+                        Default::default(),
+                    )),
+                    client: TestClient {
+                        channel: Arc::new(Mutex::new(recv)),
+                        paths: paths.unwrap_or_default(),
+                    },
+                    io: StdIo,
+                },
+                send,
+            )
+        }
+    }
+    pub struct TestClient {
+        channel: Arc<Mutex<UnboundedReceiver<Option<EncodedResponseQuery>>>>,
+        paths: HashSet<String>,
+    }
+
+    #[cfg_attr(feature = "async-send", async_trait::async_trait)]
+    #[cfg_attr(not(feature = "async-send"), async_trait::async_trait(?Send))]
+    impl Client for TestClient {
+        type Error = std::io::Error;
+
+        async fn request(
+            &self,
+            path: String,
+            _: Option<Vec<u8>>,
+            _: Option<namada_core::chain::BlockHeight>,
+            _: bool,
+        ) -> Result<EncodedResponseQuery, Self::Error> {
+            if !self.paths.contains(&path) {
+                return Err(std::io::Error::other("oh noes"));
```

Comment:
nit: maybe better to panic on these since is tests only

Reply:
ah nvm I see you're actually checking this err msg (or the one below, not sure tbh) in the tests